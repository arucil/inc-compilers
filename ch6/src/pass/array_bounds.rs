use ast::{Error, Exp, ExpKind, Program, Type};
use id_arena::Arena;
use support::Range;

pub fn insert_bounds_check(
  prog: Program<String, Type>,
) -> Program<String, Type> {
  let types = &prog.types;
  Program {
    body: prog
      .body
      .into_iter()
      .map(|exp| exp_insert(exp, types))
      .collect(),
    ..prog
  }
}

fn exp_insert(
  exp: Exp<String, Type>,
  types: &Arena<Type>,
) -> Exp<String, Type> {
  match exp.kind {
    ExpKind::Int(_)
    | ExpKind::Var(_)
    | ExpKind::Str(_)
    | ExpKind::Bool(_)
    | ExpKind::Get(_)
    | ExpKind::Void
    | ExpKind::NewLine => exp,
    ExpKind::Prim {
      op: op @ (_, "make-vector"),
      mut args,
    } => {
      let range = exp.range;
      let ty = exp.ty;
      let val = exp_insert(args.pop().unwrap(), types);
      let len = exp_insert(args.pop().unwrap(), types);
      make_check_exp(
        range,
        ty.clone(),
        len,
        |len_var| Exp {
          kind: ExpKind::Prim {
            op: (range, ">="),
            args: vec![
              len_var,
              Exp {
                kind: ExpKind::Int(0),
                range,
                ty: Type::Int,
              },
            ],
          },
          range,
          ty: Type::Bool,
        },
        |len_var| Exp {
          kind: ExpKind::Prim {
            op,
            args: vec![len_var, val],
          },
          range,
          ty,
        },
        |len_var| Error::Length(box len_var),
      )
    }
    ExpKind::Prim {
      op: op @ (_, "vector-ref"),
      mut args,
    } if matches!(args[0].ty.resolved(types), Type::Array(_)) => {
      let range = exp.range;
      let ty = exp.ty;
      let idx = exp_insert(args.pop().unwrap(), types);
      let vec = exp_insert(args.pop().unwrap(), types);
      let vec_range = vec.range;
      let tmp = "(bounds-tmp-vec)".to_owned();
      let vec_var = Exp {
        kind: ExpKind::Var(tmp.clone()),
        range: vec_range,
        ty: vec.ty.clone(),
      };
      let check_exp = make_check_exp(
        range,
        ty.clone(),
        idx,
        |idx_var| Exp {
          kind: ExpKind::Prim {
            op: (range, "and"),
            args: vec![
              Exp {
                kind: ExpKind::Prim {
                  op: (range, ">="),
                  args: vec![
                    idx_var.clone(),
                    Exp {
                      kind: ExpKind::Int(0),
                      range,
                      ty: Type::Int,
                    },
                  ],
                },
                range,
                ty: Type::Bool,
              },
              Exp {
                kind: ExpKind::Prim {
                  op: (range, "<"),
                  args: vec![
                    idx_var,
                    Exp {
                      kind: ExpKind::Prim {
                        op: (range, "vector-length"),
                        args: vec![vec_var.clone()],
                      },
                      range,
                      ty: Type::Int,
                    },
                  ],
                },
                range,
                ty: Type::Bool,
              },
            ],
          },
          range,
          ty: Type::Bool,
        },
        |idx_var| Exp {
          kind: ExpKind::Prim {
            op,
            args: vec![vec_var.clone(), idx_var],
          },
          range,
          ty: ty.clone(),
        },
        |idx_var| Error::OutOfBounds {
          index: box idx_var,
          len: box Exp {
            kind: ExpKind::Prim {
              op: (range, "vector-length"),
              args: vec![vec_var.clone()],
            },
            range,
            ty: Type::Int,
          },
        },
      );
      Exp {
        kind: ExpKind::Let {
          var: (vec_range, tmp),
          init: box vec,
          body: box check_exp,
        },
        range,
        ty,
      }
    }
    ExpKind::Prim {
      op: op @ (_, "vector-set!"),
      mut args,
    } if matches!(args[0].ty.resolved(types), Type::Array(_)) => {
      let range = exp.range;
      let ty = exp.ty;
      let val = exp_insert(args.pop().unwrap(), types);
      let idx = exp_insert(args.pop().unwrap(), types);
      let vec = exp_insert(args.pop().unwrap(), types);
      let vec_range = vec.range;
      let tmp = "(bounds-tmp-vec)".to_owned();
      let vec_var = Exp {
        kind: ExpKind::Var(tmp.clone()),
        range: vec_range,
        ty: vec.ty.clone(),
      };
      let check_exp = make_check_exp(
        range,
        ty.clone(),
        idx,
        |idx_var| Exp {
          kind: ExpKind::Prim {
            op: (range, "and"),
            args: vec![
              Exp {
                kind: ExpKind::Prim {
                  op: (range, ">="),
                  args: vec![
                    idx_var.clone(),
                    Exp {
                      kind: ExpKind::Int(0),
                      range,
                      ty: Type::Int,
                    },
                  ],
                },
                range,
                ty: Type::Bool,
              },
              Exp {
                kind: ExpKind::Prim {
                  op: (range, "<"),
                  args: vec![
                    idx_var,
                    Exp {
                      kind: ExpKind::Prim {
                        op: (range, "vector-length"),
                        args: vec![vec_var.clone()],
                      },
                      range,
                      ty: Type::Int,
                    },
                  ],
                },
                range,
                ty: Type::Bool,
              },
            ],
          },
          range,
          ty: Type::Bool,
        },
        |idx_var| Exp {
          kind: ExpKind::Prim {
            op,
            args: vec![vec_var.clone(), idx_var, val],
          },
          range,
          ty: ty.clone(),
        },
        |idx_var| Error::OutOfBounds {
          index: box idx_var,
          len: box Exp {
            kind: ExpKind::Prim {
              op: (range, "vector-length"),
              args: vec![vec_var.clone()],
            },
            range,
            ty: Type::Int,
          },
        },
      );
      Exp {
        kind: ExpKind::Let {
          var: (vec_range, tmp),
          init: box vec,
          body: box check_exp,
        },
        range,
        ty,
      }
    }
    ExpKind::Prim { op, args } => Exp {
      kind: ExpKind::Prim {
        op,
        args: args.into_iter().map(|exp| exp_insert(exp, types)).collect(),
      },
      ..exp
    },
    ExpKind::Call { name, args } => Exp {
      kind: ExpKind::Call {
        name,
        args: args.into_iter().map(|exp| exp_insert(exp, types)).collect(),
      },
      ..exp
    },
    ExpKind::Let { var, init, body } => Exp {
      kind: ExpKind::Let {
        var,
        init: box exp_insert(*init, types),
        body: box exp_insert(*body, types),
      },
      ..exp
    },
    ExpKind::If { cond, conseq, alt } => Exp {
      kind: ExpKind::If {
        cond: box exp_insert(*cond, types),
        conseq: box exp_insert(*conseq, types),
        alt: box exp_insert(*alt, types),
      },
      ..exp
    },
    ExpKind::Set { var, exp: exp1 } => Exp {
      kind: ExpKind::Set {
        var,
        exp: box exp_insert(*exp1, types),
      },
      ..exp
    },
    ExpKind::Begin { seq, last } => Exp {
      kind: ExpKind::Begin {
        seq: seq.into_iter().map(|exp| exp_insert(exp, types)).collect(),
        last: box exp_insert(*last, types),
      },
      ..exp
    },
    ExpKind::While { cond, body } => Exp {
      kind: ExpKind::While {
        cond: box exp_insert(*cond, types),
        body: box exp_insert(*body, types),
      },
      ..exp
    },
    ExpKind::Print(args) => Exp {
      kind: ExpKind::Print(
        args.into_iter().map(|exp| exp_insert(exp, types)).collect(),
      ),
      ..exp
    },
    ExpKind::Error(_) => unreachable!()
  }
}

fn make_check_exp<F, G, E>(
  range: Range,
  ty: Type,
  index: Exp<String, Type>,
  cond: F,
  conseq: G,
  err: E,
) -> Exp<String, Type>
where
  F: FnOnce(Exp<String, Type>) -> Exp<String, Type>,
  G: FnOnce(Exp<String, Type>) -> Exp<String, Type>,
  E: FnOnce(Exp<String, Type>) -> Error<String, Type>,
{
  let tmp = "(bounds-tmp-index)".to_owned();
  let idx_var = Exp {
    kind: ExpKind::Var(tmp.clone()),
    range: index.range,
    ty: Type::Int,
  };
  Exp {
    kind: ExpKind::Let {
      var: (index.range, tmp),
      init: box index,
      body: box Exp {
        kind: ExpKind::If {
          cond: box cond(idx_var.clone()),
          conseq: box conseq(idx_var.clone()),
          alt: box Exp {
            kind: ExpKind::Error(err(idx_var)),
            range,
            ty: ty.clone(),
          },
        },
        range,
        ty: ty.clone(),
      },
    },
    range,
    ty,
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use ast::*;
  use ch5::pass::typecheck::typecheck;
  use insta::assert_snapshot;

  #[test]
  fn array() {
    let prog = parse(
      r#"
(let ([x (make-vector 3 #t)]
      [i 1]
      [y (vector-ref x i)]
      [z (vector 1 #t "abc")])
  (vector-set! x i (vector-ref z 1))
  (vector-set! z 1 (not y)))
      "#,
    )
    .unwrap();
    let prog = typecheck(prog).unwrap();
    let result = insert_bounds_check(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
