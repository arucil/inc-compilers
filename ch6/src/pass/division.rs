use ast::{Error, Exp, ExpKind, Program, Type};
use id_arena::Arena;
use support::Range;

pub fn insert_division_check(
  prog: Program<String, Type>,
) -> Program<String, Type> {
  let types = &prog.types;
  Program {
    body: prog
      .body
      .into_iter()
      .map(|exp| exp_insert(exp, types))
      .collect(),
    ..prog
  }
}