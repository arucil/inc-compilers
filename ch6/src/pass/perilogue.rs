use super::register_allocation::Info;
use asm::{Arg, Block, Instr, Label, Program};

pub fn add_perilogue(mut prog: Program<Info>) -> Program<Info> {
  add_prologue(&mut prog);
  add_epilogue(&mut prog);
  prog
}

fn add_prologue(prog: &mut Program<Info>) {
  use asm::Reg::*;
  use Instr::*;
  let stack_space =
    prog.info.stack_space + prog.info.used_callee_saved_regs.len() * 8;
  let stack_space =
    ((stack_space + 15) & !15) - prog.info.used_callee_saved_regs.len() * 8;
  let mut code = vec![
    Push(Arg::Reg(Rbp)),
    Mov {
      src: Arg::Reg(Rsp),
      dest: Arg::Reg(Rbp),
    },
    Sub {
      src: Arg::Imm(stack_space as i64),
      dest: Arg::Reg(Rsp),
    },
  ];
  for &reg in &prog.info.used_callee_saved_regs {
    code.push(Push(Arg::Reg(reg)));
  }
  code.extend_from_slice(&[
    Mov {
      src: Arg::Imm(65536),
      dest: Arg::Reg(Rdi),
    },
    Mov {
      src: Arg::Imm(64),
      dest: Arg::Reg(Rsi),
    },
    Call {
      label: "rt_initialize".to_owned(),
      arity: 2,
    },
    Mov {
      src: Arg::Label("rt_rootstack_begin".to_owned()),
      dest: Arg::Reg(R15),
    },
  ]);
  for i in (0..prog.info.rootstack_space as i32).step_by(8) {
    code.push(Mov {
      src: Arg::Imm(0),
      dest: Arg::Deref(R15, i),
    });
  }
  code.extend_from_slice(&[
    Add {
      src: Arg::Imm(prog.info.rootstack_space as i64),
      dest: Arg::Reg(R15),
    },
    Jmp(Label::Start),
  ]);
  let block = Block { global: true, code };
  prog.blocks.push((Label::EntryPoint, block));
}

fn add_epilogue(prog: &mut Program<Info>) {
  use asm::Reg::*;
  use Instr::*;
  let mut code: Vec<Instr> = vec![Sub {
    src: Arg::Imm(prog.info.rootstack_space as i64),
    dest: Arg::Reg(R15),
  }];
  code.extend(
    prog
      .info
      .used_callee_saved_regs
      .iter()
      .rev()
      .map(|&reg| Pop(Arg::Reg(reg))),
  );
  code.extend_from_slice(&[
    Mov {
      src: Arg::Reg(Rbp),
      dest: Arg::Reg(Rsp),
    },
    Pop(Arg::Reg(Rbp)),
    Mov {
      src: Arg::Imm(60),
      dest: Arg::Reg(Rax),
    },
    Mov {
      src: Arg::Imm(0),
      dest: Arg::Reg(Rdi),
    },
    Syscall,
  ]);
  let block = Block {
    global: false,
    code,
  };
  prog.blocks.push((Label::Conclusion, block));
}
