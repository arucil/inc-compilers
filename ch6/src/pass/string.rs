use ast::{Error, Exp, ExpKind, FunDef, IdxVar, Program, Type};

pub fn expose_string_concat(
  prog: Program<IdxVar, Type>,
) -> Program<IdxVar, Type> {
  let mut state = State { tmp_counter: 0 };
  Program {
    fun_defs: prog
      .fun_defs
      .into_iter()
      .map(|(name, fun)| {
        (
          name,
          FunDef {
            body: state.exp_insert(fun.body),
            ..fun
          },
        )
      })
      .collect(),
    body: state.exp_insert(prog.body),
    ..prog
  }
}

struct State {
  tmp_counter: usize,
}

impl State {
  /// (string-append e1 e2)
  /// =>
  /// (let ([tmp1 e1]
  ///       [tmp2 e2]
  ///       [len1 (string-length tmp1)]
  ///       [result (alloc-string
  ///                 (+ len1 (string-length tmp2)))])
  ///   (copy-string! result 0 tmp1)
  ///   (copy-string! result len1 tmp2))
  fn exp_insert(&mut self, exp: Exp<IdxVar, Type>) -> Exp<IdxVar, Type> {
    match exp.kind {
      ExpKind::Int(_)
      | ExpKind::Var(_)
      | ExpKind::Str(_)
      | ExpKind::Bool(_)
      | ExpKind::Get(_)
      | ExpKind::Void
      | ExpKind::NewLine => exp,
      ExpKind::Prim {
        op: (_, "string-append"),
        mut args,
      } => {
        let range = exp.range;
        let arg2 = self.exp_insert(args.pop().unwrap());
        let arg1 = self.exp_insert(args.pop().unwrap());
        let tmp1 = self.new_tmp();
        let arg1_var = Exp {
          kind: ExpKind::Var(tmp1.clone()),
          range: exp.range,
          ty: Type::Int,
        };
        let tmp2 = self.new_tmp();
        let arg2_var = Exp {
          kind: ExpKind::Var(tmp2.clone()),
          range: exp.range,
          ty: Type::Int,
        };
        let len1 = self.new_tmp();
        let len1_var = Exp {
          kind: ExpKind::Var(len1.clone()),
          range: exp.range,
          ty: Type::Int,
        };
        let tmp3 = self.new_tmp();
        let result_var = Exp {
          kind: ExpKind::Var(tmp3.clone()),
          range: exp.range,
          ty: Type::Int,
        };
        let body = Exp {
          kind: ExpKind::Begin {
            seq: vec![
              Exp {
                kind: ExpKind::Prim {
                  op: (range, "copy-string!"),
                  args: vec![
                    result_var.clone(),
                    Exp {
                      kind: ExpKind::Int(0),
                      range,
                      ty: Type::Int,
                    },
                    arg1_var.clone(),
                  ],
                },
                range,
                ty: Type::Void,
              },
              Exp {
                kind: ExpKind::Prim {
                  op: (range, "copy-string!"),
                  args: vec![
                    result_var.clone(),
                    len1_var.clone(),
                    arg2_var.clone(),
                  ],
                },
                range,
                ty: Type::Void,
              },
            ],
            last: box result_var,
          },
          range,
          ty: Type::Str,
        };
        vec![
          (tmp1, box arg1),
          (tmp2, box arg2),
          (
            len1,
            box Exp {
              kind: ExpKind::Prim {
                op: (range, "string-length"),
                args: vec![arg1_var],
              },
              range,
              ty: Type::Int,
            },
          ),
          (
            tmp3,
            box Exp {
              kind: ExpKind::Prim {
                op: (exp.range, "alloc-string"),
                args: vec![Exp {
                  kind: ExpKind::Prim {
                    op: (exp.range, "+"),
                    args: vec![
                      len1_var,
                      Exp {
                        kind: ExpKind::Prim {
                          op: (range, "string-length"),
                          args: vec![arg2_var],
                        },
                        range,
                        ty: Type::Int,
                      },
                    ],
                  },
                  range: exp.range,
                  ty: Type::Int,
                }],
              },
              range,
              ty: Type::Str,
            },
          ),
        ]
        .into_iter()
        .rfold(body, |body, (var, init)| Exp {
          kind: ExpKind::Let {
            var: (range, var),
            init,
            body: box body,
          },
          range,
          ty: Type::Str,
        })
      }
      ExpKind::Prim { op, args } => Exp {
        kind: ExpKind::Prim {
          op,
          args: args.into_iter().map(|exp| self.exp_insert(exp)).collect(),
        },
        ..exp
      },
      ExpKind::Apply {
        fun,
        args,
        r#struct,
      } => Exp {
        kind: ExpKind::Apply {
          fun: box self.exp_insert(*fun),
          args: args.into_iter().map(|exp| self.exp_insert(exp)).collect(),
          r#struct,
        },
        ..exp
      },
      ExpKind::Let { var, init, body } => Exp {
        kind: ExpKind::Let {
          var,
          init: box self.exp_insert(*init),
          body: box self.exp_insert(*body),
        },
        ..exp
      },
      ExpKind::If { cond, conseq, alt } => Exp {
        kind: ExpKind::If {
          cond: box self.exp_insert(*cond),
          conseq: box self.exp_insert(*conseq),
          alt: box self.exp_insert(*alt),
        },
        ..exp
      },
      ExpKind::Set { var, exp: exp1 } => Exp {
        kind: ExpKind::Set {
          var,
          exp: box self.exp_insert(*exp1),
        },
        ..exp
      },
      ExpKind::Begin { seq, last } => Exp {
        kind: ExpKind::Begin {
          seq: seq.into_iter().map(|exp| self.exp_insert(exp)).collect(),
          last: box self.exp_insert(*last),
        },
        ..exp
      },
      ExpKind::While { cond, body } => Exp {
        kind: ExpKind::While {
          cond: box self.exp_insert(*cond),
          body: box self.exp_insert(*body),
        },
        ..exp
      },
      ExpKind::Print(args) => Exp {
        kind: ExpKind::Print(
          args.into_iter().map(|exp| self.exp_insert(exp)).collect(),
        ),
        ..exp
      },
      ExpKind::Error(Error::Length(arg)) => Exp {
        kind: ExpKind::Error(Error::Length(box self.exp_insert(*arg))),
        ..exp
      },
      ExpKind::Error(Error::OutOfBounds { index, len }) => Exp {
        kind: ExpKind::Error(Error::OutOfBounds {
          index: box self.exp_insert(*index),
          len: box self.exp_insert(*len),
        }),
        ..exp
      },
      ExpKind::Error(Error::DivByZero) => exp,
      ExpKind::FunRef { .. } => exp,
    }
  }

  fn new_tmp(&mut self) -> IdxVar {
    self.tmp_counter += 1;
    IdxVar {
      name: "(str-tmp)".to_owned(),
      index: self.tmp_counter,
    }
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use ast::*;
  use ch4::pass::uniquify;
  use ch5::pass::typecheck::typecheck;
  use insta::assert_snapshot;

  #[test]
  fn string_append() {
    let prog = parse(
      r#"
(let ([x "\n"])
  (print (string-append "abc" (string-append "+" x))))
      "#,
    )
    .unwrap();
    let prog = typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let result = expose_string_concat(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
