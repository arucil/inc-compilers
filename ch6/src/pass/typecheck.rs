#[cfg(test)]
mod tests {
  use ast::*;
  use ch5::pass::typecheck::typecheck;
  use support::{CompileError, Range};

  #[test]
  fn vector_ref0() {
    let prog =
      parse(r#"(let ([x (vector 1 "a" #t)]) (vector-ref x 0))"#).unwrap();
    let result = typecheck(prog);
    assert_eq!(
      result.map(|_| ()),
      Err(CompileError {
        range: Range { start: 0, end: 46 },
        message: "expected Void, found Int".to_owned(),
      })
    );
  }

  #[test]
  fn vector_ref1() {
    let prog =
      parse(r#"(let ([x (vector 1 "a" #t)]) (vector-ref x 1))"#).unwrap();
    let result = typecheck(prog);
    assert_eq!(
      result.map(|_| ()),
      Err(CompileError {
        range: Range { start: 0, end: 46 },
        message: "expected Void, found Str".to_owned(),
      })
    );
  }

  #[test]
  fn nested_vector_ref() {
    let prog =
      parse(r#"(let ([x (vector 1 (vector (void)))]) (vector-ref (vector-ref x 1) 0))"#).unwrap();
    let result = typecheck(prog);
    assert_eq!(result.map(|_| ()), Ok(()));
  }

  #[test]
  fn vector_len() {
    let prog = parse(r#"(let ([x (vector 1 #f)]) (vector-length x))"#).unwrap();
    let result = typecheck(prog);
    assert_eq!(
      result.map(|_| ()),
      Err(CompileError {
        range: Range { start: 0, end: 43 },
        message: "expected Void, found Int".to_owned(),
      })
    );
  }

  #[test]
  fn vector_set() {
    let prog = parse(
      r#"(let ([x (vector 1 (vector 0 "abc"))]) (vector-set! x 1 (vector 7 "")))"#,
    )
    .unwrap();
    let result = typecheck(prog);
    assert_eq!(result.map(|_| ()), Ok(()));
  }
}
