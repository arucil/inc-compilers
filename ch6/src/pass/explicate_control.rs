#[cfg(test)]
mod tests {
  use ch2::pass::remove_complex_operands;
  use ch4::pass::explicate_control;
  use ch4::pass::shrink;
  use ch4::pass::uniquify;
  use ch5::pass::typecheck;
  use insta::assert_snapshot;

  #[test]
  fn vector() {
    let prog = ast::parse(
      r#"
(let ([x (void)] [y (vector 1 #f (void) (vector (set! x x)) "abc")])
  (set! x (vector-set! y 1 (not (vector-ref y 1))))
  (set! x (vector-ref y 2))
  (vector-set! y 0 (vector-length y))
  (vector-set! y 2 (vector-ref (vector-ref y 3) 0)))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control::explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn r#struct() {
    let prog = ast::parse(
      r#"
(define-struct my-void (x Void))
(define-struct my-comp (x Int) (y Bool) (z Void) (u my-void) (v Str))
(let ([x (void)] [y (my-comp 1 #f (void) (my-void (set! x x)) "abc")])
  (set! x (set-my-comp-y! y (not (my-comp-y y))))
  (set! x (my-comp-z y))
  (set-my-comp-x! y (vector-length y))
  (set-my-comp-z! y (my-void-x (my-comp-u y))))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control::explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
