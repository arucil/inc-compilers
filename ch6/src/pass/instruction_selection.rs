#[cfg(test)]
mod tests {
  use ch2::pass::remove_complex_operands;
  use ch4::pass::explicate_control;
  use ch4::pass::instruction_selection;
  use ch4::pass::uniquify;
  use ch5::pass::typecheck;
  use insta::assert_snapshot;

  #[test]
  fn vector() {
    let prog = ast::parse(
      r#"
(let ([x (void)] [y (vector 1 #f (void) (vector (set! x x)) "abc")])
  (set! x (vector-set! y 1 (not (vector-ref y 1))))
  (set! x (vector-ref y 2))
  (vector-set! y 0 (vector-length y))
  (vector-set! y 2 (vector-ref (vector-ref y 3) 0)))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let prog = explicate_control::explicate_control(prog);
    let result = instruction_selection::select_instruction(prog, true);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn empty_struct() {
    let prog = ast::parse(
      r#"
(define-struct foo)
(let ([x (foo)])
  (void))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let prog = explicate_control::explicate_control(prog);
    let result = instruction_selection::select_instruction(prog, true);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn array() {
    let prog = ast::parse(
      r#"
(let ([x (make-vector 3 #t)]
      [y (make-vector 0 2)]
      [z (make-vector 2 x)]
      [t (make-vector 10 (void))]
      [u (void)])
  (vector-set! x 1 #f)
  (vector-set! t 3 (void))
  (set! u (vector-ref t (vector-ref y 0))))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let prog = explicate_control::explicate_control(prog);
    let result = instruction_selection::select_instruction(prog, true);
    assert_snapshot!(result.to_string_pretty());
  }
}
