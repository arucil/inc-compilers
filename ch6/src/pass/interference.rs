#[cfg(test)]
mod tests {
  use asm::{Label, Program};
  use ast::{IdxVar, Type};
  use ch3::location_set::LocationSet;
  use ch4::pass::instruction_selection::Info as OldOldInfo;
  use ch4::pass::interference;
  use ch4::pass::liveness_analysis;
  use indexmap::indexmap;
  use insta::assert_snapshot;
  use maplit::hashmap;

  #[test]
  fn vector() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      pop rdi
      pop rsi
      push x
      mov w, rbx
      call rt_allocate, 2, gc
      add w, x
      jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexmap! {
          IdxVar::new("x") => Type::Tuple(vec![Type::Int]),
          IdxVar::new("w") => Type::Int,
        },
      },
      blocks,
      ..Program::default()
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = interference::build_interference(prog);

    assert_snapshot!(format!("{}", result.info));
  }

  #[test]
  fn str() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      pop rdi
      pop rsi
      push x
      mov w, rbx
      call rt_allocate, 2, gc
      add w, x
      jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexmap! {
          IdxVar::new("x") => Type::Str,
          IdxVar::new("w") => Type::Int,
        },
      },
      blocks,
      ..Program::default()
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = interference::build_interference(prog);

    assert_snapshot!(format!("{}", result.info));
  }
}
