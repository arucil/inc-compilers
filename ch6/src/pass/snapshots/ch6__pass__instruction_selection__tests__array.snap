---
source: ch6/src/pass/instruction_selection.rs
expression: result.to_string_pretty()
---
locals: {x.1: Array(Bool), (bounds-tmp-index).0: Int, y.3: Array(Int), (bounds-tmp-index).2: Int, z.5: Array(Array(Bool)), (bounds-tmp-index).4: Int, t.7: Array(Void), (bounds-tmp-index).6: Int, (bounds-tmp-vec).9: Array(Bool), (bounds-tmp-index).10: Int, tmp.0: Int, tmp.1: Int, (bounds-tmp-vec).11: Array(Void), (bounds-tmp-index).12: Int, tmp.2: Int, tmp.3: Int, (bounds-tmp-vec).13: Array(Void), (bounds-tmp-index).16: Int, (bounds-tmp-vec).14: Array(Int), (bounds-tmp-index).15: Int, tmp.4: Int, tmp.5: Int, tmp.6: Int, tmp.7: Int, tmp.8: Int}

constants: {}

start:
    mov (bounds-tmp-index).0, 3
    cmp (bounds-tmp-index).0, 0
    jge block24
    jmp block25
block24:
    mov rsi, (bounds-tmp-index).0
    shl rsi, 3
    mov rdi, (bounds-tmp-index).0
    shl rdi, 5
    or rdi, 5
    mov rdx, r15
    call rt_allocate
    mov rdi, rax
    mov rsi, 1
    call rt_fill_array
    mov x.1, rax
    jmp block23
block25:
    mov rdi, (bounds-tmp-index).0
    call rt_length_error
block23:
    mov (bounds-tmp-index).2, 0
    cmp (bounds-tmp-index).2, 0
    jge block21
    jmp block22
block21:
    mov rsi, (bounds-tmp-index).2
    shl rsi, 3
    mov rdi, (bounds-tmp-index).2
    shl rdi, 5
    or rdi, 5
    mov rdx, r15
    call rt_allocate
    mov rdi, rax
    mov rsi, 2
    call rt_fill_array
    mov y.3, rax
    jmp block20
block22:
    mov rdi, (bounds-tmp-index).2
    call rt_length_error
block20:
    mov (bounds-tmp-index).4, 2
    cmp (bounds-tmp-index).4, 0
    jge block18
    jmp block19
block18:
    mov rsi, (bounds-tmp-index).4
    shl rsi, 3
    mov rdi, (bounds-tmp-index).4
    shl rdi, 5
    or rdi, 13
    mov rdx, r15
    call rt_allocate
    mov rdi, rax
    mov rsi, x.1
    call rt_fill_array
    mov z.5, rax
    jmp block17
block19:
    mov rdi, (bounds-tmp-index).4
    call rt_length_error
block17:
    mov (bounds-tmp-index).6, 10
    cmp (bounds-tmp-index).6, 0
    jge block15
    jmp block16
block15:
    mov rsi, 0
    mov rdi, (bounds-tmp-index).6
    shl rdi, 5
    or rdi, 21
    mov rdx, r15
    call rt_allocate
    mov t.7, rax
    jmp block14
block16:
    mov rdi, (bounds-tmp-index).6
    call rt_length_error
block14:
    mov (bounds-tmp-vec).9, x.1
    mov (bounds-tmp-index).10, 1
    cmp (bounds-tmp-index).10, 0
    jge block13
    jmp block12
block13:
    mov r11, (bounds-tmp-vec).9
    mov tmp.0, qword [r11]
    shr tmp.0, 5
    cmp (bounds-tmp-index).10, tmp.0
    jl block11
    jmp block12
block12:
    mov r11, (bounds-tmp-vec).9
    mov tmp.1, qword [r11]
    shr tmp.1, 5
    mov rdi, (bounds-tmp-index).10
    mov rsi, tmp.1
    call rt_out_of_bounds_error
block11:
    mov r11, (bounds-tmp-index).10
    add r11, 1
    shl r11, 3
    add r11, (bounds-tmp-vec).9
    mov qword [r11], 0
    jmp block10
block10:
    mov (bounds-tmp-vec).11, t.7
    mov (bounds-tmp-index).12, 3
    cmp (bounds-tmp-index).12, 0
    jge block9
    jmp block8
block9:
    mov r11, (bounds-tmp-vec).11
    mov tmp.2, qword [r11]
    shr tmp.2, 5
    cmp (bounds-tmp-index).12, tmp.2
    jl block7
    jmp block8
block8:
    mov r11, (bounds-tmp-vec).11
    mov tmp.3, qword [r11]
    shr tmp.3, 5
    mov rdi, (bounds-tmp-index).12
    mov rsi, tmp.3
    call rt_out_of_bounds_error
block7:
    mov (bounds-tmp-vec).13, t.7
    mov (bounds-tmp-vec).14, y.3
    mov (bounds-tmp-index).15, 0
    cmp (bounds-tmp-index).15, 0
    jge block6
    jmp block5
block6:
    mov r11, (bounds-tmp-vec).14
    mov tmp.4, qword [r11]
    shr tmp.4, 5
    cmp (bounds-tmp-index).15, tmp.4
    jl block4
    jmp block5
block5:
    mov r11, (bounds-tmp-vec).14
    mov tmp.5, qword [r11]
    shr tmp.5, 5
    mov rdi, (bounds-tmp-index).15
    mov rsi, tmp.5
    call rt_out_of_bounds_error
block4:
    mov r11, (bounds-tmp-index).15
    add r11, 1
    shl r11, 3
    add r11, (bounds-tmp-vec).14
    mov (bounds-tmp-index).16, qword [r11]
    jmp block3
block3:
    cmp (bounds-tmp-index).16, 0
    jge block2
    jmp block1
block2:
    mov r11, (bounds-tmp-vec).13
    mov tmp.6, qword [r11]
    shr tmp.6, 5
    cmp (bounds-tmp-index).16, tmp.6
    jl block0
    jmp block1
block1:
    mov r11, (bounds-tmp-vec).13
    mov tmp.7, qword [r11]
    shr tmp.7, 5
    mov rdi, (bounds-tmp-index).16
    mov rsi, tmp.7
    call rt_out_of_bounds_error
block0:
    mov r11, t.7
    mov tmp.8, qword [r11]
    shr tmp.8, 5
    mov rdi, tmp.8
    call rt_print_int
    call rt_print_newline
    jmp conclusion

